### PR TITLE
Null safety, Add Action2

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Motivation
We need to make w_flux null safe.

w_flux's Action class specifically has some issues when migrating to null safety that need special attention and migration.

Due to Action.call having an optional paramater, it must be typed as nullable. However, this means that we can't make listener payloads non-nullable, since there's no guarantee that the argument was specified.

```
class Action<T> /*...*/ {
  Future call([T? payload]) { 
    for (final listener in _listeners) {
      await listener(payload);
      //             ^^^^^^^
      // Error: can't assign T? to T
    }
  }
  
  ActionSubscription listen(dynamic onData(T event)) {/*...*/}
  
  /*...*/
}
```
As a result, we have to make the listener argument nullable:

  ActionSubscription listen(dynamic onData(T? event)) {/*...*/}
To be able to support non-nullable payloads (in addition to nullable payloads), we'll want to make a new Action2 class with required payloads, and migrate consumers over to it before those consumers migrate to null safety.

## Changes
- Migrate w_flux to null safety
- Create Action2 class that supports non-nullable payloads (in addition to nullable payloads)
- Add tests  coverage for new Action2 class

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Architecture member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/w_flux/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/w_flux/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: example/pubspec.yaml [hunk 1]
```diff
@@ -22,3 +21,7 @@ dev_dependencies:
 dependency_overrides:
   w_flux:
     path: ../
+  react:
+    git:
+      url: https://github.com/Workiva/react-dart.git
+      ref: v7_wip
```
Comment (sorenthompson-wk) [inline example/pubspec.yaml:27]: I'm not sure what the plan is for maintaining this. Are we ok with leaving this directed to the v7 branch for now?

File: test/action_test.dart [hunk 2]
```diff
@@ -199,18 +197,233 @@ void main() {
...
+        voidAction.listen((_) {
+          c.complete();
+        });
+
+        await voidAction(null);
+        return c.future;
+      });
+    });
+
+    group('benchmarks', () {
```
Comment (sorenthompson-wk) [inline test/action_test.dart:397]: This test was copied from the original Action class and is skipped. I'm not sure why. When un-skipped it passes.

Should we un-skip?

**Comment by evanweible-wf** [comment]
Closing this - we'll be creating a new PR for the null safety migration now that ActionV2 has landed in master.

**Pull request closed without merging**